### PR TITLE
no_color: Fix coloring bug in --no-color option

### PR DESCRIPTION
**Pull request by shreyans800755**
Fixes https://github.com/coala/coala/issues/4185

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by shreyans800755** [comment]
I'm not really sure how to check the color of the log being printed in stdout. I've added a test case to get the coverage to 100%.

**Comment by shreyans800755** [comment]
@Makman2 @thisch 
Can someone please review this ?

File: coalib/output/Logging.py [hunk 1]
```diff
@@ -41,37 +41,64 @@ def configure_logging():
...
+                    'class': 'coalib.output.Logging.CounterHandler'
+                }
+            },
+            'root': {
+                'level': 'DEBUG',
+                'handlers': ['plain', 'counter']
+            },
+            'formatters': {
+                'plain': {
+                    '()': 'colorlog.ColoredFormatter',
```
Comment (Makman2) [inline coalib/output/Logging.py]: is there no plain formatter?

File: coalib/output/Logging.py [hunk 2]
```diff
@@ -41,37 +41,64 @@ def configure_logging():
...
     CounterHandler.reset()
 
-    logging.config.dictConfig({
-        'version': 1,
-        'handlers': {
-            'colored': {
-                'class': 'logging.StreamHandler',
-                'formatter': 'colored',
-                'stream': sys.stderr
+    if not no_color:
```
Comment (Makman2) [inline coalib/output/Logging.py]: this looks too complicated to build up two completely different dicts just because of this param. Can't you modify only the entries of the current dict that would be affected by `no-color = True`? As far as I see we have to add the `plain` handler and formatter to the config and just use it for `root.handlers` instead of the `colored`-handler.

If you really don't want to create the color formatters if `no-color` is given, you should try to build upon a dict with common settings. In this case only the version is common, but at least we don't duplicate information:
```
config = {'version': 1}

if not_color:
    # I don't know a better name now for this variable...
    print_config = ...
else:
    print_config = ...

config.update(print_config)
logging.config.dictConfig(config)
```

File: coalib/output/Logging.py [hunk 3]
```diff
@@ -32,7 +32,7 @@ def get_num_calls_for_level(cls, level):
         return cls._call_counter[level]
 
 
-def configure_logging():
+def configure_logging(no_color=False):
```
Comment (Makman2) [inline coalib/output/Logging.py]: why not `color` instead of `no_color`? This looks like doing things the other way around ;D

Comment (Makman2) [inline coalib/output/Logging.py]: Although coala uses a `no-color` argument, imo the internal API doesn't have to follow the nomenclature 1:1 ;)

**Comment by shreyans800755** [comment]
@Makman2 Rebased.

File: coalib/output/Logging.py [hunk 4]
```diff
@@ -41,37 +41,46 @@ def configure_logging():
...
-    logging.config.dictConfig({
-        'version': 1,
-        'handlers': {
-            'colored': {
-                'class': 'logging.StreamHandler',
-                'formatter': 'colored',
-                'stream': sys.stderr
+    formatter = 'colored' if color else 'plain'
+
+    if color is True:
```
Comment (Makman2) [inline coalib/output/Logging.py]: and if `color == False`? Then you don't configure logging at all :O

Comment (shreyans800755) [inline coalib/output/Logging.py]: Ouch. Forgot to remove the condition.

File: coalib/output/Logging.py [hunk 5]
```diff
@@ -60,15 +62,21 @@ def configure_logging():
...
                 'datefmt': '%X',
                 'log_colors': {
                     'ERROR': 'red',
                     'WARNING': 'yellow',
                     'INFO': 'blue',
                     'DEBUG': 'green'
                 }
+            },
+            'plain': {
+                '()': 'colorlog.ColoredFormatter',
```
Comment (mrtes) [inline coalib/output/Logging.py]: I believe you can remove this line, and it will use the default formatter of the logging package.
The default should be able to handle the format string as well, no need to use a custom one in this case.

Comment (shreyans800755) [inline coalib/output/Logging.py]: Makes sense. Updated.

**Comment by shreyans800755** [comment]
LLVM page is not accessible by appveyor, and not even in my local for that matter. It is causing appveyor build failure.

File: coalib/output/Logging.py [hunk 6]
```diff
@@ -41,12 +41,14 @@ def configure_logging():
     # reset counter handler
     CounterHandler.reset()
 
+    formatter = 'colored' if color else 'plain'
```
Comment (Makman2) [inline coalib/output/Logging.py]: if you want you can inline that into the dict, but up to you ;)

File: tests/coalaTest.py [hunk 7]
```diff
@@ -179,3 +179,19 @@ def test_logged_error_causes_non_zero_exitcode(self):
...
+                prepare_file(['#fixme'], None) as (lines, filename):
+            retval, stdout, stderr = execute_coala(
+                             coala.main,
+                             'coala', '-c', os.devnull,
+                             '-f', re.escape(filename),
+                             '-b', 'LineCountTestBear')
+            self.assertIn('This file has 1 lines.',
+                          stdout,
+                          'The output should report count as 1 lines')
+            self.assertIn('During execution of coala', stderr)
```
Comment (Makman2) [inline tests/coalaTest.py]: Why does an error happen?

**Comment by Makman2** [comment]
Even I can't access the download of LLVM... Filing an issue. This needs to be fixed before merge :/

**Comment by Makman2** [comment]
Filed here: https://github.com/coala/coala/issues/4246

File: coalib/output/Logging.py [hunk 8]
```diff
@@ -60,15 +62,19 @@ def configure_logging():
         'formatters': {
             'colored': {
                 '()': 'colorlog.ColoredFormatter',
-                'format': '%(log_color)s[%(levelname)s]%(reset)s[%(asctime)s]'
-                          ' %(message)s',
+                'format': '%(log_color)s[%(levelname)s]%(reset)s'
+                          '[%(asctime)s] %(message)s',
```
Comment (Makman2) [inline coalib/output/Logging.py]: hmm if possible remove the change or make a different commit, it's a layout change ;) Also the previous layout is valid, so it's more personal preference ;)

Comment (shreyans800755) [inline coalib/output/Logging.py]: This is just to make line length <80. I didn't intend to change the layout.
Will change it back to how it was.

File: tests/coalaTest.py (hunk 7 above)
Comment (shreyans800755) [inline tests/coalaTest.py]: Not sure. I've just copied test case `test_coala` from https://github.com/coala/coala/blob/master/tests/coalaTest.py#L26. And made changes related to what is needed to be tested for the changes I made in source.

Comment (Makman2) [inline tests/coalaTest.py]: Ah okay then it's alright. I just wondered why something is reported as an error (or maybe warning) :)

File: tests/coalaTest.py [hunk 9]
```diff
@@ -179,3 +179,19 @@ def test_logged_error_causes_non_zero_exitcode(self):
...
+        with bear_test_module(), \
+                prepare_file(['#fixme'], None) as (lines, filename):
+            retval, stdout, stderr = execute_coala(
+                             coala.main,
+                             'coala', '-c', os.devnull,
+                             '-f', re.escape(filename),
+                             '-b', 'LineCountTestBear')
+            self.assertIn('This file has 1 lines.',
+                          stdout,
+                          'The output should report count as 1 lines')
```
Comment (Makman2) [inline tests/coalaTest.py]: Up to you:
Maybe we want to be a bit more restrictive here? We know that the non-colored output has no special characters so we may assert for the complete message. But I would allow this too^^

File: tests/coalaTest.py (hunk 7 above)
Comment (mrtes) [inline tests/coalaTest.py]: All the logging stuff is configured to be logged to stderr.

File: tests/coalaTest.py (hunk 9 above)
Comment (shreyans800755) [inline tests/coalaTest.py]: I'm making stderr check more restrictive. I think timestamps are also there which can't be checked easily. IMO stdout check is good enough.

**Comment by Makman2** [comment]
Gaaaah and now language-check... anyway going to ack :+1:

**Comment by Makman2** [comment]
ack 7cf63bb

**Comment by Makman2** [comment]
Btw filed an issue for the unknown source of the stderr error message: https://github.com/coala/coala/issues/4249

**Comment by shreyans800755** [comment]
> Btw filed an issue for the unknown source of the stderr error message: #4249
That looks good.

**Comment by Makman2** [comment]
ack 123424c

**Comment by Makman2** [comment]
@rultor merge

**Comment by Makman2** [comment]
Let's retry...
@rultor merge

**Comment by Makman2** [comment]
@rultor merge

**Comment by Makman2** [comment]
Still not working... let's wait a while, your PR is approved anyway ;)
(hopefully no conflicts arise meanwhile :3)

**Comment by shreyans800755** [comment]
Ah! Same error again. Isn't there any other way to merge PR ?
@Makman2 
We shouldn't rely on a single system for such an importance task, I guess.

**Comment by Makman2** [comment]
Merging is a single process, so we have a single system for it ;) rultor is sometimes not so reliable all the time. But don't worry we don't lose much time, we cherry pick yours and other approved PRs into a master PR as soon as merging flows again.

Maybe we get someday our own merge bot... with more features :3 (and no damn security holes...)

File: tests/coalaTest.py [hunk 10]
```diff
@@ -179,3 +179,23 @@ def test_logged_error_causes_non_zero_exitcode(self):
...
+            self.assertIn('An unknown error occurred. This is a bug. '
```
Comment (jayvdb) [inline tests/coalaTest.py]: Why is this allowing stderr with `This is a bug. ... This should never happen.`
Why is there an unknown error?  Why is that a good thing?
If it really is a good thing, it definitely needs a comment to explain why this should be in the output.

Comment (shreyans800755) [inline tests/coalaTest.py]: I've copied this test case from https://github.com/coala/coala/blob/master/tests/coalaTest.py#L26 and made changes into it related to issue.

Comment (jayvdb) [inline tests/coalaTest.py]: Found the problem, and it is fixable by improving the test logic that you copied.  See https://github.com/coala/coala/pull/4266

**Comment by jayvdb** [comment]
unack 123424c

**Comment by shreyans800755** [comment]
@Makman2 @jayvdb Updated test case to check coloring functionality properly.
Not sure why `circleCI` is failing.
As per logs, pytest is timed out.

> ========== 1 failed, 723 passed, 1 pytest-warnings in 111.75 seconds ===========

And as per https://github.com/coala/coala/blob/master/circle.yml#L23, timeout is 900 sec.
Why is test failing with timeout before 900 sec ? Or 900 sec is for overall CI action ?

Surprisingly rest of the CI are passing fine.

File: tests/coalaTest.py [hunk 11]
```diff
@@ -179,3 +179,24 @@ def test_logged_error_causes_non_zero_exitcode(self):
...
+                prepare_file(['#fixme'], None) as (lines, filename):
+            retval, stdout, stderr = execute_coala(
+                coala.main, 'coala')
+            errors = filter(bool, stderr.split('\n'))
+            # Every error message must start with characters
+            # used for coloring.
+            for err in errors:
+                self.assertNotRegex(err, '^\[WARNING\]')
+            self.assertEqual(
+                retval, 0, 'coala must return zero when there are no errors')
```
Comment (Makman2) [inline tests/coalaTest.py]: could you move the color testing into an own testing function `test_coala_color`?
(Hm maybe we should call then both functions `test_coala_log_with_colors` and `test_coala_log_without_colors` as `test_coala_color` sounds kind of bad :D)

**Comment by Makman2** [comment]
Let's just retry for CircleCI, maybe it's a random fail^^

**Comment by jayvdb** [comment]
The commit message prefix "no_color" is rather unusual.  I would pick "Logging.py" as that is the main part of the codebase that is being modified.

**Comment by shreyans800755** [comment]
> Let's just retry for CircleCI, maybe it's a random fail^^
@Makman2 You're right.
@jayvdb Updated.

**Comment by Makman2** [comment]
ack ff52cab

**Comment by Makman2** [comment]
oops needs a rebase again :3

**Comment by shreyans800755** [comment]
@Makman2 Rebased.

**Comment by Makman2** [comment]
reack e05c816

**Comment by Makman2** [comment]
@rultor merge
